### PR TITLE
Feature/201 citests4caas

### PR DESCRIPTION
**Pull request by ycliuhw**
----

## Description of change

Initialise the steps to write CI test for caas

## Documentation changes

Does not affect current user workflow

### CODE REVIEW

File: acceptancetests/assess_caas_bootstrap.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+
+__metaclass__ = type
+
+
+log = logging.getLogger("assess_caas_bootstrap")
+
+
+def assess_caas_bootstrap(client):
+    # Deploy charms, there are several under ./repository
```
Comment (wallyworld) [inline acceptancetests/assess_caas_bootstrap.py]: I don't think this comment is true - we're not deploying charms in here but bootstrapping a k8s cluster

File: acceptancetests/assess_caas_bootstrap.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    repository_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'repository')
+    bundle = local_charm_path(charm='bundles-kubernetes-core-lxd.yaml',
+                              repository=repository_dir, juju_ver=client.version)
+    client.deploy_bundle(bundle, static_bundle=True)
+    # Wait for the deployment to finish.
+    client.wait_for_started()
+    client.set_config(service='kubernetes-worker', options={"proxy-extra-args": "proxy-mode=userspace"})
```
Comment (wallyworld) [inline acceptancetests/assess_caas_bootstrap.py]: We could put this into the bundle

File: acceptancetests/winazurearm.py [hunk 3]
```diff
@@ -292,7 +292,7 @@ def delete_instance(client, name_id, resource_group=None):
         if deployment:
             log.debug('Found {} {}'.format(resource_group.name, deployment.name))
             if not client.read_only:
-                poller = rgd.delete_vm(deployment.name)
+                poller = rgd.delete_vm(deployment.name)  # noqa ?? ./winazurearm.py:295:26: F821 undefined name 'rgd'
```
Comment (wallyworld) [inline acceptancetests/winazurearm.py]: what's this?

File: acceptancetests/assess_caas_bootstrap.py (hunk 1 above)
Comment (ycliuhw) [reply acceptancetests/assess_caas_bootstrap.py]: yes, there is a lint target in `Makefile` now, It will need to be added to the jenkins stage

File: acceptancetests/winazurearm.py (hunk 3 above)
Comment (ycliuhw) [reply acceptancetests/winazurearm.py]: ah, i think I can add lint to test to ensure running `lint` with `test` together. if we do run test, it will be good.

File: acceptancetests/Makefile [hunk 4]
```diff
@@ -1,31 +1,73 @@
-mkfile_dir := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
```
Comment (Veebers) [inline acceptancetests/Makefile]: This Makefile change should be in a PR of it's own as it's not specific to the addition of the caas functional testing.

File: acceptancetests/jujupy/client.py [hunk 5]
```diff
@@ -2121,6 +2122,19 @@ def switch(self, model=None, controller=None):
...
+        self.juju('scp', args)
+
+
+class K8sClient(ModelClient):
+    ...
+
+
+class CaasClient(ModelClient):
+    ...
+
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: What is expected to be contained in these 2 classes? Does it make sense to subclass ModelClient?

Is a K8sClient an extension of a juju model? Same question for CaasClient. How does a user interact with a CaasClient of K8sClient. 

I suspect a CaasClient is very much like a ModelClient, what makes then different? Are they interchangable?

File: acceptancetests/assess_caas_bootstrap.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+__metaclass__ = type
+
+
+log = logging.getLogger("assess_caas_bootstrap")
+
+
+def assess_caas_bootstrap(client):
+    # Deploy k8s bundle to spin up k8s cluster
+    repository_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'repository')
```
Comment (Veebers) [inline acceptancetests/assess_caas_bootstrap.py]: The local repository dir should be provided by the env var 'JUJU_REPOSITORY' (you'll see local_charm_path looks for it if repository is not passed.)

This seems a little opaque, but it does separate the location of the repo from the test into the testing environment (i.e. the CI test jobs define JUJU_REPOSITORY).

File: acceptancetests/assess_caas_bootstrap.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    bundle = local_charm_path(charm='bundles-kubernetes-core-lxd.yaml',
+                              repository=repository_dir, juju_ver=client.version)
+    client.deploy_bundle(bundle, static_bundle=True)
+    # Wait for the deployment to finish.
+    client.wait_for_started()
```
Comment (Veebers) [inline acceptancetests/assess_caas_bootstrap.py]: You'll probably need to wait for workloads too, perhaps there might be something more you'll need to wait for? Perhaps workloads is enough.

wait_for_started and wait_for_workloads are the default 'wait for' methods used, but BaseCondition allows one to write specific "wait for this status within a timeout". See jujupy.wait_condition.UnitInstallCondition as an example of a specific status check.

File: acceptancetests/jujupy/client.py [hunk 8]
```diff
@@ -2121,6 +2122,19 @@ def switch(self, model=None, controller=None):
             raise ValueError('No target to switch to has been given.')
         self.juju('switch', (':'.join(args),), include_e=False)
 
+    def scp(self, _from, _to):
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: A couple things:
  - Why do from and to have leading underscores? (they're not 'private' variables etc.)
  - "-m .." is not needed as ModelClient.juju() will add that to the command arguments (unless include_e is False).

I'm not sure there is value adding scp() to ModelClient (seeing as the -m addition is not needed). This could be done in the test as client.juju('scp', (to, from)). It would also need to expose '--no-host-key-checks' and '--proxy' options to be complete.

File: acceptancetests/assess_caas_bootstrap.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    # ensure kube credentials
+    client.scp(_from='kubernetes-master/0:config', _to='~/.kube/config')
+
+    # TODO:
```
Comment (Veebers) [inline acceptancetests/assess_caas_bootstrap.py]: Can you expand on what these steps will entail? This would help shape what k8sclient and CaasClient might look like.

File: acceptancetests/jujupy/client.py (hunk 8 above)
Comment (ycliuhw) [reply acceptancetests/jujupy/client.py]: 1. `from` can not be used as var name, so we usually added `_` like `_from` or `from_`, so keep consistency, I added `_` for `to` as well.
2. good point, if we are under current client, `model name` is not needed.

File: acceptancetests/assess_caas_bootstrap.py (hunk 6 above)
Comment (ycliuhw) [reply acceptancetests/assess_caas_bootstrap.py]: you are right, we should use the `JUJU_REPOSITORY` rather than finding it from relative path.

**Comment by ycliuhw** [comment]
[adding CI test coverage for caas deployments](https://trello.com/c/ufM5qXUs/201-adding-ci-test-coverage-for-caas-deployments)

File: acceptancetests/Makefile (hunk 4 above)
Comment (ycliuhw) [reply acceptancetests/Makefile]: yeah, can do. I will raise another PR for `Makefile` change later

**Comment by ycliuhw** [comment]
$$build$$

**Comment by ycliuhw** [comment]
!!build!!

File: acceptancetests/jujupy/client.py [hunk 10]
```diff
@@ -2121,6 +2128,90 @@ def switch(self, model=None, controller=None):
...
+        self.juju('scp', (_from, _to))
+
+
+# caas `add-k8s` did not implement parsing kube config path via flag,
+# so parse it via env var ->  https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/loader.go#L44
+KUBE_CONFIG_PATH_ENV_VAR = 'KUBECONFIG'
+
+
+class CaasClient(ModelClient):
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: I'm not convinced this is the right way to achieve this split.

It adds complexity with the Client type selection, and it's possible there are edge cases of ```ModelClient``` creation that aren't covered here.

Perhaps we use composition instead. Something like this:
```python
class CaasClient:
    cloud_name = '...''
    ....

    def __init__(self, client):
        """:param client: ModelClient that is wrapping us."""
        self._client = client

    def add_cloud(self):
        self._client.juju('add-k8s', self.cloud_name)


# The existing ModelClient.
class ModelClient:

    def __init__(self, ...):
        self.caas = CaasClient(self)


client = ModelClient(...)
client.caas.add_cloud()
```

Thoughts?

File: acceptancetests/assess_caas_bootstrap.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python
+""" Test caas k8s cluster bootstrap
+"""
```
Comment (Veebers) [inline acceptancetests/assess_caas_bootstrap.py]: This should give a brief outline what it's actually asserting, "Spinning up a k8s cluster and ensuring that x is correct."

File: acceptancetests/jujupy/client.py [hunk 12]
```diff
@@ -2121,6 +2128,90 @@ def switch(self, model=None, controller=None):
...
+    kube_config_path = None
+
+    def _ensure_dir(self, path):
+        try:
+            os.makedirs(path)
+        except OSError:
+            if not os.path.isdir(path):
+                raise
+
+    def __prepare(self):
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: This should just be in the ```__init__```.

File: acceptancetests/jujupy/client.py [hunk 13]
```diff
@@ -2121,6 +2128,90 @@ def switch(self, model=None, controller=None):
...
+
+
+class CaasClient(ModelClient):
+
+    cloud_name = 'k8cloud'
+    kubectl_path = None
+    kube_home = None
+    kube_config_path = None
+
+    def _ensure_dir(self, path):
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: Is ```jujupy.utility.ensure_dir``` suitable?

File: acceptancetests/jujupy/client.py [hunk 14]
```diff
@@ -2121,6 +2128,90 @@ def switch(self, model=None, controller=None):
...
+
+    def cluster_up(self, bundle_path):
+        self.__prepare()
+
+        self.deploy_bundle(bundle_path, static_bundle=True)
+        # Wait for the deployment to finish.
+        self.wait_for_started()
+
+        # ensure kube home
+        self._ensure_dir(self.kube_home)
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: This should be part of __prepare (which should be in ```__init__```)

File: acceptancetests/jujupy/client.py [hunk 15]
```diff
@@ -2121,6 +2128,90 @@ def switch(self, model=None, controller=None):
...
+        try:
+            cluster_info = self._kubectl_cmd('cluster-info')
+            log.debug('cluster_info -> \n%s', cluster_info)
+            nodes_info = self._kubectl_cmd('get nodes')
+            log.debug('nodes_info -> \n%s', nodes_info)
+            return True
+        except RuntimeError:
+            return False
+
+    def _kubectl_cmd(self, args):
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: kubectl is tied pretty close to caas? Should this be a 'private' method or perhaps just 'def kubectl(...)' much like juju(...) us for ModelClient.

This would expose the kubectl command (assuming it's needed) so a test author could do something like (assuming the suggestion above re: moving CaasClient within ModelClient):
```python

client.caas.kubectl('some', 'command')
```

File: acceptancetests/jujupy/client.py [hunk 16]
```diff
@@ -2262,7 +2353,24 @@ def _get_full_path(juju_path):
         return os.path.abspath(juju_path)
 
 
-def client_from_config(config, juju_path, debug=False, soft_deadline=None):
+def decide_client(client_type):
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: Moving ```CaasClient``` within ```ModelClient``` removes a lot of this complexity.

File: acceptancetests/jujupy/utility.py [hunk 17]
```diff
@@ -17,7 +17,7 @@
 from contextlib import contextmanager
```
Comment (Veebers) [inline acceptancetests/jujupy/utility.py]: Are these changes needed for the PR?

File: acceptancetests/assess_caas_bootstrap.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+__metaclass__ = type
+
+
+log = logging.getLogger("assess_caas_bootstrap")
+
+
+def assess_caas_bootstrap(client):
+    # Deploy k8s bundle to spin up k8s cluster
+    bundle = local_charm_path(charm='bundles-kubernetes-core-lxd.yaml',
+                              repository=os.environ['JUJU_REPOSITORY'], juju_ver=client.version)
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply acceptancetests/assess_caas_bootstrap.py]: Here I have to parse in `repository` or it will go to `./repository/subfolders/*` rather than `./repository`.

File: acceptancetests/jujupy/client.py [hunk 19]
```diff
@@ -851,7 +858,7 @@ def add_model(self, env):
             env = self.env.clone(env)
         model_client = self.clone(env)
         with model_client._bootstrap_config() as config_file:
-            self._add_model(env.environment, config_file)
+            self._add_model(env.environment, config_file, cloud_region=None)
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply acceptancetests/jujupy/client.py]: ah, u r right

File: acceptancetests/jujupy/client.py (hunk 10 above)
Comment (ycliuhw) [reply acceptancetests/jujupy/client.py]: I think `CaasClient` and `ModelClient` are two similar but different, or you can think `Caas` model is a `special` model. So ModelClient should not have anything related Caas, but Caas should have all the attr/abilities of ModelClient and has some other stuff that ModelClient does `not` have.

File: acceptancetests/jujupy/client.py (hunk 12 above)
Comment (ycliuhw) [reply acceptancetests/jujupy/client.py]: it was initially in `__init__` but it didnot work. because `juju_home` is `JUJU_HOME=/tmp/test-run` in contructor `__init__`, but in `context_manager`, `juju_home` was switched to `juju_home/juju-homes/<current-build>` which is what I need

File: acceptancetests/jujupy/client.py (hunk 14 above)
Comment (ycliuhw) [reply acceptancetests/jujupy/client.py]: the reason I put it here is the below two `scp` is the reason for ensuring `kube_home`, so I prefer to put related code together.

File: acceptancetests/jujupy/client.py (hunk 15 above)
Comment (ycliuhw) [reply acceptancetests/jujupy/client.py]: I was trying to make the cmd generic, so we will need to wrap related actions(like `is_cluster_healthy`) here rather than calling kubectl from assess script.

File: acceptancetests/jujupy/client.py (hunk 16 above)
Comment (ycliuhw) [reply acceptancetests/jujupy/client.py]: I think we have already had different clients, I suggest we should refactor `ClientType.Fake` related code as well.

File: acceptancetests/assess_caas_bootstrap.py (hunk 11 above)
Comment (ycliuhw) [reply acceptancetests/assess_caas_bootstrap.py]: agreed

File: acceptancetests/assess_caas_bootstrap.py [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+
+    k8s_model = client.add_caas_model('testcaas')  # noqa
+    if not client.is_cluster_healthy:
+        raise JujuAssertionError('k8s cluster is not healthy coz kubectl is not accessible')
+
+
+def parse_args(argv):
+    """Parse all arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--client_type', default=ClientType.Caas)
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply acceptancetests/assess_caas_bootstrap.py]: Here, we only need to add this flag when we are writing tests for caas, it's not required for other normal models

File: acceptancetests/jujupy/client.py (hunk 8 above)
Comment (ycliuhw) [reply acceptancetests/jujupy/client.py]: removed

File: acceptancetests/jujupy/wait_condition.py [hunk 21]
```diff
@@ -22,13 +22,13 @@
     VersionsNotUpdated,
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply acceptancetests/jujupy/wait_condition.py]: i think here is broken.

File: acceptancetests/jujupy/client.py [hunk 22]
```diff
@@ -2122,6 +2133,80 @@ def switch(self, model=None, controller=None):
...
+        args = args if isinstance(args, list) else [args]
+        args = [sub_flag for flag in args for sub_flag in flag.split(' ') if sub_flag]
+        args = [self.kubectl_path, '--kubeconfig={}'.format(self.kube_config_path)] + args
+        return self._sh(args)
+
+    def _sh(self, cmd):
+        cmd = cmd if isinstance(cmd, list) else [cmd]
+        cmd_str = ' '.join(cmd)
+        log.debug('_sh: cmd -> `%s`, \n\tcmd_splitted -> %s', cmd, cmd_str)
+        exitcode, data = subprocess.getstatusoutput(cmd_str)
```
Comment (ycliuhw) [inline acceptancetests/jujupy/client.py]: sorry, @Veebers  `getstatusoutput` will only work for `py3`, 
Are we going to run this job in py3 or py2?

**Comment by ycliuhw** [comment]
Suggestions (https://paste.ubuntu.com/p/wYmmjMF8nw/) for design of KubeClient from @Veebers , put here for later reference as IRC flushing quickly, thx

File: acceptancetests/deploy_stack.py [hunk 23]
```diff
@@ -110,6 +111,16 @@ def deploy_dummy_stack(client, charm_series, use_charmstore=False):
         client.wait_for_started(3600)
 
 
+def deploy_caas_stack(bundle_path, client):
+
```
Comment (Veebers) [inline acceptancetests/deploy_stack.py]: newline not needed.

File: acceptancetests/deploy_stack.py [hunk 24]
```diff
@@ -110,6 +111,16 @@ def deploy_dummy_stack(client, charm_series, use_charmstore=False):
...
+    client.deploy_bundle(bundle_path, static_bundle=True)
+
```
Comment (Veebers) [inline acceptancetests/deploy_stack.py]: newline not needed.

File: acceptancetests/jujupy/client.py [hunk 25]
```diff
@@ -2122,6 +2124,61 @@ def switch(self, model=None, controller=None):
         self.juju('switch', (':'.join(args),), include_e=False)
 
 
+# caas `add-k8s` did not implement parsing kube config path via flag,
+# so parse it via env var ->  https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/loader.go#L44
+KUBE_CONFIG_PATH_ENV_VAR = 'KUBECONFIG'
+
+
+class CaasClient(object):
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: "(object)" isn't needed due to ```__metaclass__ = type```, this should just be: ```class CaasClient:"

File: acceptancetests/jujupy/client.py [hunk 26]
```diff
@@ -2122,6 +2124,61 @@ def switch(self, model=None, controller=None):
...
+            return True
+        except subprocess.CalledProcessError as e:
+            log.error('error -> %s', e)
+            return False
+
+    def kubectl(self, *args):
+        args = (self.kubectl_path, '--kubeconfig', self.kube_config_path) + args
+        return self._sh(args)
+
+    def _sh(self, args):
```
Comment (Veebers) [inline acceptancetests/jujupy/client.py]: is this method needed, seems like it should just be in kubectl.

File: acceptancetests/assess_caas_bootstrap.py [hunk 27]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
+""" Test caas k8s cluster bootstrap
+
+    Spining up k8s cluster and asserting the cluster is `healthy`
```
Comment (Veebers) [inline acceptancetests/assess_caas_bootstrap.py]: It would be nice to quickly outline what 'healthy' means in this context.
Also typo, spining -> spinning.

File: acceptancetests/assess_caas_bootstrap.py [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+    # Deploy k8s bundle to spin up k8s cluster
+    bundle = local_charm_path(
+        charm='bundles-kubernetes-core-lxd.yaml',
+        repository=os.environ['JUJU_REPOSITORY'],
+        juju_ver=client.version
+    )
+
+    caas_client = deploy_caas_stack(bundle_path=bundle, client=client)
+
+    k8s_model = caas_client.add_model('testcaas')  # noqa
```
Comment (Veebers) [inline acceptancetests/assess_caas_bootstrap.py]: if we don't care about the model we could just drop it on the floor and remove the noqa.

File: acceptancetests/assess_caas_bootstrap.py [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+    if not caas_client.is_cluster_healthy:
+        raise JujuAssertionError('k8s cluster is not healthy coz kubectl is not accessible')
+
+
+def parse_args(argv):
+    """Parse all arguments."""
+    parser = argparse.ArgumentParser()
```
Comment (Veebers) [inline acceptancetests/assess_caas_bootstrap.py]: This needs a description.

**Comment by ycliuhw** [comment]
&&merge&&

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[write initial CI tests for CAAS](https://trello.com/c/hzpnaWNR/205-write-initial-ci-tests-for-caas)
